File: Document_Legacy/Coding_Problem_Solving/Immigration.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <climits>
using namespace std;
using LL = long long;

LL answer = 0;
LL totalPerson;

LL GetSumAndUpdateAns(LL time, vector<int>& vi) {
	LL sum = accumulate(vi.begin(), vi.end(), (LL)0, [time](LL l, LL r) {return l + time / r; });
	if (sum >= totalPerson) {
		answer = min(answer, time);
	}
	return sum;
}

long long solution(int n, vector<int> times) {

	totalPerson = n;
	sort(times.begin(), times.end());
	answer = (LL)times.back() * n;
	LL low = 0, high = (LL)times.back() * n, mid;


	while (high - low > 1) {
		mid = (high + low) / 2;
		if (GetSumAndUpdateAns(mid, times) >= totalPerson) {
			high = mid;
		}
		else {
			low = mid;
		}
	}

	for (LL i = -1; i <= 1; i++) {
		GetSumAndUpdateAns(low + i, times);
	}
	return answer;
}

File: Document_Legacy/Coding_Problem_Solving/Immigration.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class Solution
{
    long answer = 0;
    long totalPerson;

    long GetSumAndUpdateAns(long time, List<long> list)
    {
        long sum = list.Aggregate((long)0, (l, r) => l + time / r);
        if (sum >= totalPerson)
        {
            answer = Math.Min(answer, time);
        }
        return sum;
    }

    public long solution(int n, int[] times)
    {
        var ts = new List<long>(times.Select((i) => (long)i));
        ts.Sort();
        totalPerson = n;
        answer = ts.Last() * (long)n;
        long low = 0, high = times.Last() * (long)n, mid;

        while (high - low > 1)
        {
            mid = (high + low) / 2;
            if (GetSumAndUpdateAns(mid, ts) >= totalPerson)
            {
                high = mid;
            }
            else
            {
                low = mid;
            }
        }
        for (long i = -1; i <= 1; i++)
        {
            GetSumAndUpdateAns(low + i, ts);
        }
        return answer;
    }
}
